### PR TITLE
feat: add `UnionExec`

### PR DESCRIPTION
**Pull request by iajoiner**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [x] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [x] I have run the ci check script with `source scripts/run_ci_checks.sh`.
- The following upstream PRs have been approved and merged:
  - [x] #405
  - [x] #423

# Rationale for this change
We need to add support for `UNION ALL`.

# What changes are included in this PR?
- add `UnionExec`
- add some code to simplify debugging of sumchecks

# Are these changes tested?

Yes.

### CODE REVIEW

File: crates/proof-of-sql/src/sql/proof_plans/union_exec.rs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+        accessor: &IndexMap<ColumnRef, S>,
+        _result: Option<&OwnedTable<S>>,
+        one_eval_map: &IndexMap<TableRef, S>,
+    ) -> Result<TableEvaluation<S>, ProofError> {
+        let input_table_evals = self
+            .inputs
+            .iter()
+            .map(|input| input.verifier_evaluate(builder, accessor, None, one_eval_map))
+            .collect::<Result<Vec<_>, _>>()?;
+        let num_parts = input_table_evals.len();
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/union_exec.rs:81]: NIT:
```suggestion
        let num_parts = self.inputs.len();
```

File: crates/proof-of-sql/src/sql/proof_plans/union_exec.rs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+    output_eval: &[S],
+    input_one_evals: &[S],
+    output_one_eval: S,
+) {
+    assert_eq!(input_evals.len(), input_one_evals.len());
+    let c_star_evals = input_evals
+        .iter()
+        .zip(input_one_evals)
+        .map(|(&input_eval, &input_one_eval)| {
+            let c_fold_eval = alpha * input_one_eval + fold_vals(beta, input_eval);
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/union_exec.rs:206]: We want to ensure that c_star_eval is actually padded with zeros. Having input_one_eval here means that it can be padded with anything the prover wants, which is not secure.
```suggestion
            let c_fold_eval = alpha + fold_vals(beta, input_eval);
```

File: crates/proof-of-sql/src/sql/proof_plans/union_exec.rs [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+            // c_fold * c_star - input_ones = 0
+            builder.produce_sumcheck_subpolynomial_evaluation(
+                &SumcheckSubpolynomialType::Identity,
+                c_fold_eval * c_star_eval - input_one_eval,
+            );
+            c_star_eval
+        })
+        .collect::<Vec<_>>();
+
+    let d_bar_fold_eval = alpha * output_one_eval + fold_vals(beta, output_eval);
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/union_exec.rs:217]: Same as above
```suggestion
    let d_bar_fold_eval = alpha + fold_vals(beta, output_eval);
```

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]

File: crates/proof-of-sql/src/sql/proof/verification_builder.rs [hunk 4]
```diff
@@ -109,6 +109,7 @@ impl<'a, S: Scalar> VerificationBuilder<'a, S> {
     }
 
     /// Produce the evaluation of a subpolynomial used in sumcheck
+    #[allow(clippy::trivially_copy_pass_by_ref)]
```
Comment (iajoiner) [inline crates/proof-of-sql/src/sql/proof/verification_builder.rs:112]: I will make that change in a separate PR

File: crates/proof-of-sql/src/base/polynomial/multilinear_extension.rs
Comment (JayWhite2357) [inline crates/proof-of-sql/src/base/polynomial/multilinear_extension.rs:1]: These should be in a different PR if we decide we want to merge them.

File: crates/proof-of-sql/src/sql/proof_plans/mod.rs [hunk 5]
```diff
@@ -36,11 +36,15 @@ pub(crate) use group_by_exec::GroupByExec;
...
 
 mod slice_exec;
-#[allow(unused_imports)]
 pub(crate) use slice_exec::SliceExec;
 #[cfg(all(test, feature = "blitzar"))]
 mod slice_exec_test;
 
+mod union_exec;
+pub(crate) use union_exec::UnionExec;
+#[cfg(all(test, feature = "blitzar"))]
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/mod.rs:45]: Let's not require `blitzar`.

File: crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        borrowed_bigint("a0", [1_i64, 2, 3, 4, 5], &alloc),
+        borrowed_varchar("b0", ["1", "2", "3", "4", "5"], &alloc),
+    ]);
+    let t0 = "sxt.t0".parse().unwrap();
+    let data1 = table([
+        borrowed_bigint("a1", [2_i64, 3, 4, 5, 6], &alloc),
+        borrowed_varchar("b1", ["2", "3", "4", "5", "6"], &alloc),
+    ]);
+    let t1 = "sxt.t1".parse().unwrap();
+    let mut accessor = TableTestAccessor::<InnerProductProof>::new_empty_with_setup(());
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs:58]: Can you use the NaiveEvaluationProof instead?

File: crates/proof-of-sql/src/sql/proof_plans/mod.rs (hunk 5 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/mod.rs:45]: This isn't possible until we can replace blitzar here https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/sql/proof/verifiable_query_result_test_utility.rs#L33 or we won't be able to exercise verification.

File: crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs (hunk 6 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs:58]: Sure but that requires a different PR to refactor https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/crates/proof-of-sql/src/sql/proof/verifiable_query_result_test_utility.rs and replace every instance of `InnerProductProof` with `NaiveEvaluationProof` in examples that need to be modified.

File: crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+            VerifiableQueryResult,
+        },
+        proof_exprs::test_utility::*,
+    },
+};
+use blitzar::proof::InnerProductProof;
+use bumpalo::Bump;
+
+#[test]
+fn we_can_prove_and_get_the_correct_empty_result_from_a_union_exec() {
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs:23]: Can you add some more edge cases:
* Union of 1 table.
* Union of 0 tables.

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]
LGTM

File: crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs (hunk 7 above)
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/union_exec_test.rs:23]: @JayWhite2357 `UnionExec` only exists when there is at least one `UNION` in a query which requires at least two tables. Hence we shouldn't do more than an `assert` which I already did to ban unions with 0 or 1 tables.

**Review by JayWhite2357 — DISMISSED** [review]

**Review by JayWhite2357 — APPROVED** [review]
